### DISCUSSION TITLE
Why constructors of embedded objects must not have required parameters?

### DISCUSSION THREAD

**Discussion by Ayad-Ben-Saliem** [isar/isar#1157] [Help]
As the question title explains, Why does Isar enforce constructors of embedded objects must not have required parameters?

Sometimes I really need to make some parameters required for embedded objects.
What is the point of making the constructors must not have required parameters?

**Comment by AmirhosseinShk** [comment]
That's my question and my problem too!

**Comment by androidwoman** [comment]
I have this problem too

**Comment by shahmirzali49** [comment]
@simc why ?

**Comment by giga10** [comment]
See https://github.com/isar/isar/issues/853#issuecomment-1307092947

**Reply by Ayad-Ben-Saliem** [reply]
But, Why?

**Reply by shilangyu** [reply]
This limitation has been lifted in isar 4

**Reply by adnanjpg** [reply]
@shilangyu that's cool, but there's no `4.x` release for `isar_generator`, the latest is `3.1.0+1`, how did you solve that?

**Reply by shilangyu** [reply]
I don't, sorry. Afaik isar 4 just isn't ready yet. Probably you could for now reference the new generator by url? (If the new generator even exists yet)

**Reply by adnanjpg** [reply]
@shilangyu hey, I've found the code of the generator to be under `packages/isar/lib/src/generator`, and added the reference to my pubspec.yaml like the following:

<img width="866" alt="Screenshot 2023-10-21 at 11 12 03 AM" src="https://github.com/isar/isar/assets/54329101/498c43de-41b0-44f6-96ee-444a4e0d3471">

but there appears to not be a `pubspec.yaml` file for the generator, so the `flutter pub get` command is failing with:

```
Could not find a file named "packages/isar/lib/src/generator/pubspec.yaml" in git@github.com:isar/isar.git 95e1f02c274bb4bb80f98c1a42ddf33f3690a50c.
exit code 1
```

**Reply by shilangyu** [reply]
It seems that now isar itself is the generator, so you shouldn't need `isar_generator` at all https://github.com/isar/isar/commit/41988e4fd6810c43d0a21fa9a6d3423b6f017916#diff-fd971c42a4e4a06528e360af4d00e85c91269b5dcb3a016f62f88b68a2527809

**Reply by adnanjpg** [reply]
@shilangyu oh seems I've missed checking on that, thank you so mcuh
